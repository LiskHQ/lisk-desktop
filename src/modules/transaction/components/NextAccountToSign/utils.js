import { useAccounts } from '@account/hooks';
import useTxInitiatorAccount from '@transaction/hooks/useTxInitiatorAccount';
import { getNextAccountToSign } from '@transaction/utils/multisignatureUtils';

export function useNextAccountToSign({ transactionJSON }) {
  const { getAccountByPublicKey } = useAccounts();

  const { txInitiatorAccount } = useTxInitiatorAccount({
    senderPublicKey: transactionJSON.senderPublicKey,
  });

  const nextAccountToSign = getNextAccountToSign({
    getAccountByPublicKey,
    transactionJSON,
    txInitiatorAccount,
  });

  return { nextAccountToSign };
}
