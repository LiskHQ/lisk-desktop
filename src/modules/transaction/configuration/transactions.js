const maxMessageLength = 64;
const MIN_ACCOUNT_BALANCE = 5000000;
const MIN_FEE_PER_BYTE = 1000;

const DUMMY_RECIPIENT_ADDRESS = 'lskdxc4ta5j43jp9ro3f8zqbxta9fn6jwzjucw7yt';
const DEFAULT_NUMBER_OF_SIGNATURES = 1;
const DEFAULT_SIGNATURE_BYTE_SIZE = 64;

const VOTE_AMOUNT_STEP = 1e9;

export {
  maxMessageLength,
  MIN_ACCOUNT_BALANCE,
  MIN_FEE_PER_BYTE,
  DEFAULT_NUMBER_OF_SIGNATURES,
  DUMMY_RECIPIENT_ADDRESS,
  DEFAULT_SIGNATURE_BYTE_SIZE,
  VOTE_AMOUNT_STEP,
};
