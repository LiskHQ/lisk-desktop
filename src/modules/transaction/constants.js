export const FEE_TYPES = {
  BYTES_FEE: 'bytesFee',
  PRIORITY_FEE: 'priorityFee',
  USER_ACCOUNT_INITIALIZATION: 'userAccountInitialization',
  ESCROW_ACCOUNT_INITIALIZATION: 'escrowAccountInitialization',
  VALIDATOR_REGISTRATION: 'validatorRegistration',
};

export const TransactionExecutionResult = {
  INVALID: -1,
  FAIL: 0,
  OK: 1,
};

export const ERROR_EVENTS = {
  insufficientFee: 'Please use custom fee and provide a higher fee to process this transaction.',
  invalidSignature:
    'Failed to process transaction due to invalid signature, please restart the transaction from scratch.',
};

export const EVENT_DATA_RESULT = {
  1: 'Failed to process this transaction due to insufficient balance.',
  2: 'Failed to process this transaction due to message too long',
  3: 'Failed to process this transaction due to invalid token ID',
  4: 'Failed to process this transaction due to token not supported.',
  5: 'Failed to process this transaction due to insufficient locked amount.',
  11: 'Failed to process this transaction due to token ID not available.',
  12: 'Failed to process this transaction due to token ID not native.',
  13: 'Failed to process this transaction due to insufficient escrow balance.',
  14: 'Failed to process this transaction due to invalid receiving chain.',
};
