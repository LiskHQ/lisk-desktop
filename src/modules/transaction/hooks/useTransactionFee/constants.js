export const FEE_TYPES = {
  BYTES_FEE: 'bytesFee',
  INITIALIZE: 'initialize',
  INITIALIZE_ESCROW: 'initializeEscrow',
  REGISTER_VALIDATOR: 'registerValidator',
};
