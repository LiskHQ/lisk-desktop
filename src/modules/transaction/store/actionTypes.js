const actionTypes = {
  emptyTransactionsData: 'EMPTY_TRANSACTIONS_DATA',
  pendingTransactionAdded: 'PENDING_TRANSACTION_ADDED',
  getTransactions: 'GET_TRANSACTIONS',
  getTransactionsSuccess: 'GET_TRANSACTIONS_SUCCESS',
  getTransactionSuccess: 'GET_TRANSACTION_SUCCESS',
  transactionAddDelegateName: 'TRANSACTION_ADD_DELEGATE_NAME',
  transactionCleared: 'TRANSACTION_CLEARED',
  transactionLoadFailed: 'TRANSACTION_LOAD_FAILED',
  transactionLoadRequested: 'TRANSACTION_LOAD_REQUESTED',
  transactionCreatedSuccess: 'TRANSACTION_CREATED_SUCCESS',
  transactionSignError: 'TRANSACTION_SIGN_ERROR',
  transactionDoubleSigned: 'TRANSACTION_DOUBLE_SIGNED',
  transactionsRetrieved: 'TRANSACTION_RETRIEVED',
  resetTransactionResult: 'RESET_TRANSACTION_RESULTS',
  broadcastedTransactionSuccess: 'BROADCAST_TRANSACTION_SUCCESS',
  broadcastedTransactionError: 'BROADCASTED_TRANSACTION_ERROR',
  updateTransactions: 'UPDATE_TRANSACTIONS',
  signatureSkipped: 'SIGNATURE_SKIPPED',
};

export default actionTypes;
