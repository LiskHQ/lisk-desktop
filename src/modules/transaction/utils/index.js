export { default as normalizeStakesForTx } from './normalizeStakesForTx';
export * from './moduleCommand';
export * from './transaction';
export * from './transactionDetailsHelper';
export * from './removeDuplicateTransactions';
