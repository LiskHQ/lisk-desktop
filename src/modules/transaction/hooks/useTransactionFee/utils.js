import { transactions } from '@liskhq/lisk-client';
import { joinModuleAndCommand } from '@transaction/utils/moduleCommand';
import { fromTransactionJSON } from '../../utils/encoding';

/**
 * This is solution uses the same logic as implemented in transactions utility.
 * We'll replace this easily with said class methods.
 *
 * @param {object} transaction transaction object in JSON format
 * @param {object} paramsSchema transaction schema as retrieved from Service
 * @param {int} numberOfSignatures number of signatures on the sender account
 * @param {array} priorities list of objects with value (number) and title (string), and selected (boolean)
 * @param {boolean} isTxValid defines if the form was validated successfully
 * @param {bigint} extraFee
 *
 * @returns {bigint} the transaction fee in Beddows
 */

export const computeTransactionMinFee = (
  transactionJSON,
  paramsSchema,
  numberOfSignatures,
  extraCommandFee
) => {
  const options = {
    numberOfSignatures,
    numberOfEmptySignatures: 0,
    additionalFee: BigInt(extraCommandFee),
  };

  let convertedTx = {};
  try {
    convertedTx = fromTransactionJSON(transactionJSON, paramsSchema);
  } catch (exp) {
    convertedTx = {};
  }

  return transactions.computeMinFee(convertedTx, paramsSchema, options);
};

export const getParamsSchema = (transaction, schemas) => {
  const moduleCommand = joinModuleAndCommand({
    module: transaction.module,
    command: transaction.command,
  });

  return schemas[moduleCommand];
};
