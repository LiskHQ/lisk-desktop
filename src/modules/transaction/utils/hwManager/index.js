import { updateTransactionSignatures } from '@wallet/utils/hwManager';
import { signTransaction } from '@libs/hwManager/communication';
/**
 * signTransactionByHW - Function.
 * This function is used for sign a send hardware wallet transaction.
 */
const signTransactionByHW = async (
  wallet,
  networkIdentifier,
  transactionObject,
  transactionBytes,
  keys
) => {
  const data = {
    deviceId: wallet.hwInfo.deviceId,
    index: wallet.hwInfo.derivationIndex,
    networkIdentifier,
    transactionBytes,
  };

  try {
    const signature = await signTransaction(data);
    return updateTransactionSignatures(wallet, transactionObject, signature, keys);
  } catch (error) {
    throw new Error(error);
  }
};

// eslint-disable-next-line import/prefer-default-export
export { signTransactionByHW };
