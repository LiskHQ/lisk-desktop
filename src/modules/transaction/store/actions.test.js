import loginTypes from 'src/modules/auth/const/loginTypes';
import * as hwManager from '@transaction/utils/hwManager';
import httpApi from 'src/utils/http';
import * as transactionUtils from '@transaction/utils/transaction';
import { getState } from '@fixtures/transactions';
import { sampleTransaction } from '@tests/constants/transactions';
import accounts from '@tests/constants/wallets';
import commonActionTypes from '@common/store/actions/actionTypes';
import actionTypes from './actionTypes';
import {
  emptyTransactionsData,
  transactionsRetrieved,
  resetTransactionResult,
  pendingTransactionAdded,
  transactionDoubleSigned,
  transactionBroadcasted,
  transactionCreated,
  multisigTransactionSigned,
  signatureSkipped,
} from './actions';

jest.mock('@dpos/validator/api');
jest.mock('@transaction/utils/hwManager');
jest.mock('@common/utilities/api/http');

describe('actions: transactions', () => {
  beforeEach(() => {
    jest.resetAllMocks();
  });

  const dispatch = jest.fn();

  describe('getTransactions', () => {
    const data = {
      address: 'lsks6uckwnap7s72ov3edddwgxab5e89t6uy8gjt6',
      limit: 20,
      offset: 0,
      filters: {},
    };

    it('should create an action function', () => {
      expect(typeof transactionsRetrieved(data)).toBe('function');
    });

    it('should dispatch getTransactionsSuccess action if resolved', async () => {
      // Arrange
      httpApi.mockResolvedValue({ data: [], meta: { total: 0 } });
      const expectedAction = {
        count: 0,
        confirmed: [],
        address: data.address,
        filters: data.filters,
        offset: 0,
      };

      // Act
      await transactionsRetrieved(data)(dispatch, getState);

      // Assert
      expect(dispatch).toHaveBeenNthCalledWith(1, {
        type: commonActionTypes.loadingStarted,
        data: actionTypes.transactionsRetrieved,
      });
      expect(dispatch).toHaveBeenNthCalledWith(2, {
        type: actionTypes.transactionsRetrieved,
        data: expectedAction,
      });
      expect(dispatch).toHaveBeenNthCalledWith(3, {
        type: commonActionTypes.loadingFinished,
        data: actionTypes.transactionsRetrieved,
      });
    });

    it('should dispatch getTransactionsSuccess action if resolved and default argument values are used', async () => {
      // Arrange
      const transactionData = {
        address: 'lsks6uckwnap7s72ov3edddwgxab5e89t6uy8gjt6',
      };
      httpApi.mockResolvedValue({
        data: transactionData,
        meta: { total: 0 },
      });
      const expectedAction = {
        count: 0,
        confirmed: transactionData,
        address: transactionData.address,
        filters: {},
        offset: 0,
      };

      // Act
      await transactionsRetrieved(transactionData)(dispatch, getState);

      // Assert
      expect(dispatch).toHaveBeenNthCalledWith(1, {
        type: commonActionTypes.loadingStarted,
        data: actionTypes.transactionsRetrieved,
      });
      expect(dispatch).toHaveBeenNthCalledWith(2, {
        type: actionTypes.transactionsRetrieved,
        data: expectedAction,
      });
      expect(dispatch).toHaveBeenNthCalledWith(3, {
        type: commonActionTypes.loadingFinished,
        data: actionTypes.transactionsRetrieved,
      });
    });

    it('should dispatch transactionsLoadFailed action if rejected', async () => {
      // Arrange
      const transactionError = new Error('Transaction retrieve error');
      httpApi.mockRejectedValue(transactionError);

      // Act
      await transactionsRetrieved(data)(dispatch, getState);

      // Assert
      expect(httpApi).rejects.toThrow(transactionError);
      expect(dispatch).toHaveBeenCalledTimes(3);
      expect(dispatch).toHaveBeenNthCalledWith(1, {
        type: commonActionTypes.loadingStarted,
        data: actionTypes.transactionsRetrieved,
      });
      expect(dispatch).toHaveBeenNthCalledWith(2, {
        type: actionTypes.transactionLoadFailed,
        data: { error: transactionError },
      });
      expect(dispatch).toHaveBeenNthCalledWith(3, {
        type: commonActionTypes.loadingFinished,
        data: actionTypes.transactionsRetrieved,
      });
    });
  });

  describe('emptyTransaction', () => {
    it('should create an action to empty transactions', () => {
      const expectedAction = { type: actionTypes.emptyTransactionsData };
      expect(emptyTransactionsData()).toEqual(expectedAction);
    });
  });

  describe('pendingTransactionAdded', () => {
    it('should create an action to add pending transaction', () => {
      // Arrange
      const data = {
        moduleAssetId: '1938573839:g45krEIjwK',
        id: '4emF3me9YJSbcIuOp',
        fee: '1032519n',
        nonce: '2n',
        signatures: ['xnVCm30IUhtYidgBX', 'uxsFGiaqS3n4ydB'],
        sender: {
          address: '3040783849904107057L',
          publicKey: 'MIICXAIBAAKBgQCqGKukO1De7zhZj6+H0qtjTkVxwTCpvKe4eCZ0FPqri0cb2JZfXJ/DgYSF6',
        },
        isPending: true,
      };
      const expectedAction = {
        type: actionTypes.pendingTransactionAdded,
        data,
      };

      // Assert
      expect(pendingTransactionAdded(data)).toEqual(expectedAction);
    });
  });

  describe('resetTransactionResult', () => {
    it('should create an action to reset transaction result', () => {
      const expectedAction = { type: actionTypes.resetTransactionResult };
      expect(resetTransactionResult()).toEqual(expectedAction);
    });
  });

  describe('transactionCreated', () => {
    const state = getState();

    const activeAccount = {
      ...state.wallet.info.LSK,
      hwInfo: {
        deviceModel: 'Ledger Nano S',
      },
      passphrase: state.wallet.passphrase,
    };
    const getStateWithHW = () => ({
      ...state,
      wallet: {
        info: {
          LSK: activeAccount,
        },
        hwInfo: {
          deviceModel: 'Ledger Nano S',
        },
        passphrase: state.wallet.passphrase,
      },
    });

    it('should dispatch transactionCreatedSuccess action if there are no errors', async () => {
      // Arrange
      const data = {
        amount: '21000000',
        data: '',
        recipientAddress: 'lsky3t7xfxbcjf5xmskrbhkmwzxpowex6eubghtws',
        fee: 141000,
      };

      // Act
      await transactionCreated(data)(dispatch, getState);

      // Assert
      expect(hwManager.signTransactionByHW).not.toHaveBeenCalled();
      // Replace toMatchSnapshot with a definitive assertion.
      expect(dispatch).toMatchSnapshot();
    });

    it('should dispatch transactionSignError action if there are errors during transaction creation', async () => {
      // Arrange
      const data = {
        amount: '21000000',
        data: '',
        recipientAddress: 'lsky3t7xfxbcjf5xmskrbhkmwzxpowex6eubghtws',
        fee: 141000,
      };
      const transactionError = new Error('Transaction create error');
      loginTypes.passphrase.code = 1;
      jest.spyOn(hwManager, 'signTransactionByHW')
        .mockRejectedValue(transactionError);
      const expectedAction = {
        type: actionTypes.transactionSignError,
        data: transactionError,
      };
      // Act
      await transactionCreated(data)(dispatch, getStateWithHW);
      expect(dispatch).toHaveBeenCalledWith(expectedAction);
    });
  });

  describe('transactionDoubleSigned', () => {
    const { network, wallet, settings } = getState();
    const getStateWithTx = () => ({
      network,
      wallet: {
        ...wallet,
        secondPassphrase: accounts.genesis.passphrase,
        info: {
          ...wallet.info,
          LSK: accounts.multiSig,
        },
      },
      settings,
      transactions: {
        signedTransaction: sampleTransaction,
      },
    });

    it('should create an action to store double signed tx', async () => {
      // Consume the utility
      await transactionDoubleSigned()(dispatch, getStateWithTx);

      // Prepare expectations
      const expectedAction = {
        type: actionTypes.transactionDoubleSigned,
        data: expect.any(Object),
      };

      // Assert
      expect(dispatch).toHaveBeenCalledWith(expectedAction);
    });

    it.skip('should create an action to store signature error', async () => {
      // Prepare the store
      const error = new Error('error signing tx');
      jest.spyOn(transactionUtils, 'sign')
        .mockImplementation(() => error);

      // Consume the utility
      await transactionDoubleSigned()(dispatch, getStateWithTx);

      // Prepare expectations
      const expectedAction = {
        type: actionTypes.transactionSignError,
        data: error,
      };

      // Assert
      expect(dispatch).toHaveBeenCalledWith(expectedAction);
    });
  });

  describe('transactionBroadcasted', () => {
    it('should dispatch broadcastedTransactionSuccess action if there are no errors', async () => {
      // Arrange
      httpApi.mockResolvedValue({ data: sampleTransaction });
      jest
        .spyOn(global, 'Date')
        .mockImplementationOnce(() => new Date('2021-09-15T11:07:29.864Z'));
      const expectedAction = {
        type: actionTypes.broadcastedTransactionSuccess,
        data: sampleTransaction,
      };
      const lastExpectedAction = {
        type: actionTypes.timerReset,
        data: new Date(),
      };

      // Act
      await transactionBroadcasted(sampleTransaction)(dispatch, getState);

      // Assert
      expect(dispatch).toHaveBeenCalledWith(expectedAction);
      expect(dispatch).toHaveBeenLastCalledWith(lastExpectedAction);
    });

    it('should dispatch broadcastedTransactionError action when broadcast has an error', async () => {
      // Arrange
      const transactionBroadcastError = new Error('Transaction broadcast error');
      httpApi.mockRejectedValue(transactionBroadcastError);
      const expectedAction = {
        type: actionTypes.broadcastedTransactionError,
        data: {
          error: transactionBroadcastError,
          transaction: sampleTransaction,
        },
      };

      // Act
      await transactionBroadcasted(sampleTransaction)(dispatch, getState);

      // Assert
      expect(httpApi).toHaveBeenCalled();
      expect(dispatch).toHaveBeenCalledWith(expectedAction);
    });

    it('should dispatch pendingTransactionAdded action if transformedTransaction sender address is the same as LSK address', async () => {
      const transformedAccountTransaction = {
        ...sampleTransaction,
        id: 'ad0e0acbe8a3ece3087c8362149ca39c470e565d268df32e57de5d3fe2e1ea5c',
      };
      // Arrange
      httpApi.mockResolvedValue({ data: sampleTransaction });
      const expectedAction1 = {
        type: actionTypes.broadcastedTransactionSuccess,
        data: transformedAccountTransaction,
      };
      // const expectedAction2 = {
      //   type: actionTypes.pendingTransactionAdded,
      //   data: { ...transformedAccountTransaction, isPending: true },
      // };
      const expectedAction3 = { data: expect.anything(), type: actionTypes.timerReset };

      // Act
      await transactionBroadcasted(sampleTransaction)(dispatch, getState);

      // Assert
      expect(httpApi).toHaveBeenCalled();
      expect(dispatch).toHaveBeenNthCalledWith(1, expectedAction1);
      // expect(dispatch).toHaveBeenNthCalledWith(2, expectedAction2);
      expect(dispatch).toHaveBeenNthCalledWith(3, expectedAction3);
    });
  });

  describe('multisigTransactionSigned', () => {
    const { network, wallet, settings } = getState();
    const getStateWithTx = () => ({
      network,
      wallet: {
        ...wallet,
        secondPassphrase: accounts.multiSig.passphrase,
        info: {
          ...wallet.info,
          LSK: accounts.multiSig,
        },
      },
      settings,
      transactions: {
        signedTransaction: sampleTransaction,
      },
    });
    const params = {
      rawTransaction: { signatures: [] },
      sender: { data: accounts.multiSig },
    };

    it('should create an action to store double signed tx', async () => {
      // Consume the utility
      jest.spyOn(transactionUtils, 'signMultisigTransaction').mockImplementation(() => [{ id: 1 }, undefined]);
      await multisigTransactionSigned(params)(dispatch, getStateWithTx);

      // Prepare expectations
      const expectedAction = {
        type: actionTypes.transactionDoubleSigned,
        data: expect.any(Object),
      };

      // Assert
      expect(dispatch).toHaveBeenCalledWith(expectedAction);
    });

    it('should create an action to store signature error', async () => {
      // Prepare the store
      const error = { message: 'error signing tx' };
      jest.spyOn(transactionUtils, 'signMultisigTransaction').mockImplementation(() => [undefined, error]);

      // Consume the utility
      await multisigTransactionSigned(params)(dispatch, getStateWithTx);

      // Prepare expectations
      const expectedAction = {
        type: actionTypes.transactionSignError,
        data: error,
      };

      // Assert
      expect(dispatch).toHaveBeenCalledWith(expectedAction);
    });
  });

  describe('signatureSkipped', () => {
    const props = {
      rawTransaction: {
        id: '9dc584c07c9d7ed77d54b6f8f43b8341c50e108cbcf582ceb3513388fe4ba84c',
        moduleAssetId: '2:0',
        fee: '10000000',
        nonce: 0,
        sender: {
          publicKey: '00f046aea2782180c51f7271249a0c107e6b6295c6b3c31e43c1a3ed644dcdeb',
        },
        asset: {
          amount: '200',
          recipient: { address: 'lskz5r2nbgwrzjctbcffyrn8k74jdxdmd9cj9ng45' },
          data: 'test',
        },
        signatures: [
          'af975f7670394a1fe7eee4785c2a65f4604b1e7ad2e4367308738c86ed4837ace17960b32d6d7ceca2d5c2ae7709465bb69581d4ef2713ccef1d4ed4618c0d0a',
          'ae67d1418797e48afae2c8d3641bd0ee4807e307d22df71a84d50555438a03c3df1f9162b2d7e54836979e78925261944e2c5aadaf9f9534b5cb8956fa95f501',
        ],
      },
    };

    it('should return the tx ready to be broadcasted', () => {
      expect(signatureSkipped(props)).toEqual(
        expect.objectContaining({ type: actionTypes.signatureSkipped }),
      );
    });
  });
});
