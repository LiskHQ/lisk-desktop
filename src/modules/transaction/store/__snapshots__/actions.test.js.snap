// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`actions: transactions transactionCreated should dispatch transactionCreatedSuccess action if there are no errors 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "data": [TypeError: Cannot read property 'serviceUrl' of undefined],
        "type": "TRANSACTION_SIGN_ERROR",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
